### DISCUSSION TITLE
Is it possible to use this to reset the device password?

### DISCUSSION THREAD

**Discussion by TheNitek** [mcw0/DahuaConsole#14] [Q&A]
Is it possible to use the DahueConsole to reset the device password? I can use it to change all my config settings now, but somehow I lost the password for my device.

**Comment by p2p-sys** [comment]
Try create new admin2 account, login from admin2 account, delete first admin account, create new admin account with new password, login from new admin account, delete admin2 account

**Reply by TheNitek** [reply]
Any hint on how to do that? I couldn't find anything to add/remove/modify users.

**Reply by p2p-sys** [reply]
For example:

**deleteUser:**
call("userManager.deleteUser", {
            "name": username,
        })

**modifyPassword:**
call("userManager.modifyPassword", {
            'name': name,
            'pwd': pwd,
            'pwdOld': pwd_old
        })

**getAuthorityList:**

.call("userManager.getAuthorityList", {})

**example:**
authority_list = ["CtrPanel","ShutDown","Monitor","Monitor_01","Monitor_02","Monitor_03","Monitor_04","Monitor_05","Monitor_06","Monitor_07","Monitor_08","Replay","Replay_01","Replay_02","Replay_03","Replay_04","Replay_05","Replay_06","Replay_07","Replay_08","Record","Backup","MHardisk","MPTZ","Account","Sysinfo","Config","QueryLog","DelLog","SysUpdate","Control","AutoMaintain","GeneralConf","EncodeConf","RecordConf","NetConf","AlarmConf","VideoConfig","PtzConfig","OutputConfig","DefaultConfig","DataFormat","bkConfig","SpotSet","TVSet","ATM_POS","ColorSet","OfflineLoginedUser"]

**addUser**
call("userManager.addUser", {
                                    "user":
                                        {
                                            "Name": username,
                                            "Password": auth_hash,
                                            "Group": group,
                                            "AuthorityList": authority_list,
                                            "Sharable": True,
                                            'Reserved': False,
                                            "Type": "Normal",
                                            "Memo": "",
                                            "MacOnly": "",
                                            'MaxMonitorChannels': 0,
                                        }})

**Reply by mcw0** [reply]
Yep, that will work in principal.

**Reply by Metallheaddy** [reply]
> **addUser**
> call("userManager.addUser", {
>                                     "user":
>                                         {
>                                             "Name": username,
>                                             "Password": auth_hash,
>                                             "Group": group,
>                                             "AuthorityList": authority_list,
>                                             "Sharable": True,
>                                             'Reserved': False,
>                                             "Type": "Normal",
>                                             "Memo": "",
>                                             "MacOnly": "",
>                                             'MaxMonitorChannels': 0,
>                                         }})

Would you be so kind to explain where and how exactly I should put these lines? 
I'm little bit new in all that stuff, but I'd like to get clear) 
Thanks)

**Comment by mcw0** [comment]
For now, my plan is to break up this (messy) code more, to make it more readable, improvements,  easy to implement new functions, so I won't add any more more stuff like Ad-Hoc, the add/modify/delete users may be in next update.

**Reply by Ishoushakov** [reply]
Hi, guys.
Has anyone managed to add a new user?/
I've already broken my head.

 ```
elif  cmd[1] == 'addUser':
            query_args = {
                "method": "userManager.addUser",
                "params": {
                    "user": {
                                    "Name": "user_name",
                                    "Password": "passwd",
                                    "Group": "admin",
                                    "AuthorityList":"",
                                    "Sharable": "true",
                                    "Reserved": "false",
                                    "Type": "normal",
                                    "Memo": "",
                                    "MacOnly": "",
                                    "MaxMonitorChannels": 0,

                },
               },
            }
```
I am grateful for any ideas.

**Reply by mcw0** [reply]
"AuthorityList" cannot be empty

**Reply by bakkano1** [reply]
Hi mcw0 would it be possible to make a youtube video on how to use the Dahua console

**Reply by Ishoushakov** [reply]
> "AuthorityList" cannot be empty
Hi, mcw0.
Thanks for the answer.
Still didn't work. What am i doing wrong? 

```
    elif cmd[1] == 'add':
            query_args = {
                "method": "userManager.addUser",
                 "params": {
                    "user": {
                        "AuthorityList": [
                            "AuthUserMag",
                            "AuthSysCfg",
                            "AuthSysInfo",
                            "AuthManuCtr",
                            "AuthBackup",
                            "AuthStoreCfg",
                            "AuthEventCfg",
                            "AuthNetCfg",
                            "AuthRmtDevice",
                            "AuthSecurity",
                            "AuthMaintence",
                            "Monitor",
                            "Monitor_01",
                            "Monitor_02",
                            "Monitor_03",
                            "Monitor_04",
                            "Monitor_05",
                            "Monitor_06",
                            "Monitor_07",
                            "Monitor_08",
                            "Monitor_09",
                            "Monitor_10",
                            "Monitor_11",
                            "Monitor_12",
                            "Monitor_13",
                            "Monitor_14",
                            "Monitor_15",
                            "Monitor_16",
                            "Replay",
                            "Replay_01",
                            "Replay_02",
                            "Replay_03",
                            "Replay_04",
                            "Replay_05",
                            "Replay_06",
                            "Replay_07",
                            "Replay_08",
                            "Replay_09",
                            "Replay_10",
                            "Replay_11",
                            "Replay_12",
                            "Replay_13",
                            "Replay_14",
                            "Replay_15",
                            "Replay_16"
                        ],
                        "Group": "admin",
                        "Id": 13,
                        "MacOnly": "",
                        "MaxMonitorChannels": 0,
                        "Memo": "admin 's account",
                        "Name": "admin2",
                        "Password": "password",

                    },
                },
            }
```
Error.

**Reply by mcw0** [reply]
"Password" needs to be MD5 hash, same format as constructed in 'dhConsole.json'.
And, I also guess you take the 'AuthorityList' from the device itself before adding to this request, as it may differ a bit.

I would do: 'userManager.getAuthorityList' first, then 'dahua_gen2_md5_hash' and sum that into something like this:

        query_args = {
            'method': 'userManager.addUser',
            'params': {
                'user':
                    {
                        'AuthorityList': authority_list, # From 'userManager.getAuthorityList'
                        'Group': 'admin',
                        'Memo': 'pwn',
                        'Name': 'pwn',
                        'Password': dh_hash, # From 'dahua_gen2_md5_hash'
                    },
            },
        }

**Reply by mcw0** [reply]
[Code from upcoming commit]

    def add_user(self, msg):

        if not self.dh.dh_realm:
            log.failure(f'[add_user] no realm exist ({self.dh.dh_realm})')
            return False
        query_args = {
            'method': 'userManager.getAuthorityList',
            'params': None,
        }
        dh_data = self.dh.send_call(query_args, error_codes=True)
        authority_list = dh_data.get('params')
        dh_hash = dahua_gen2_md5_hash(
            dh_realm=self.dh.dh_realm, username='pwn', password='pwn', return_hash=True
        )

        query_args = {
            'method': 'userManager.addUser',
            'params': {
                'user':
                    {
                        'AuthorityList': authority_list,
                        'Group': 'admin',
                        'Memo': 'pwn',
                        'Name': 'pwn',
                        'Password': dh_hash,
                    },
            },
        }
        dh_data = self.dh.send_call(query_args, error_codes=True)
        if dh_data.get('result'):
            log.success('Success')
        else:
            log.failure('Failed: User already exist?')

**Reply by mcw0** [reply]
> Hi mcw0 would it be possible to make a youtube video on how to use the Dahua console

Hi,
I seriously think I'll never do a youtube video on this (I'm not youtuber), however, what I am currently doing is to split the code, rewrite some code and later stage document the code so it would be easier to understand, use it easier, make/add own functions easier... etc.

**Reply by Ishoushakov** [reply]
> dh_realm=self.dh.dh_realm, username='pwn', password='pwn', return_hash=True

Hello, mcw0
Thank you for the answer, unfortunately, I just saw it.

I add at dahua.py:
`from pwdmanager import dahua_gen2_md5_hash`

and 
```
def add_user(self, msg):

    if not self.dh.dh_realm:
        log.failure(f'[add_user] no realm exist ({self.dh.dh_realm})')
        return False
        ...........

```
Error:

```
if not self.dh.dh_realm:
AttributeError: 'DahuaFunctions' object has no attribute 'dh'
```
It looks like the last step is left. But I couldn't find a solution. 
The question is at the kindergarten level, i'm sorry for that.

Best regards.

**Reply by infiniband2** [reply]
I try modify this:
            query_args = {
                "method": "userManager.getAuthorityList",
                "params": "None",
            }
            dh_data = self.send_call(query_args, errorcodes=True)
            authority_list = dh_data.get('params')
            
            dh_hash = dahua_gen2_md5_hash(
                dh_realm='None', username='user1', password='1Qaz2Wsx3Edc', return_hash=True
            )
    
            query_args = {
                "method": "userManager.addUser",
                "params": {
                    "user": {
                        "Name": "user1",
                        "Password": dh_hash,
                        "Group": "admin",
                        #"AuthorityList": authority_list,
                        "AuthorityList": ["Monitor", "Monitor_01", "Monitor_02", "Monitor_03", "Monitor_04", "Monitor_05", "Monitor_06", "Monitor_07", "Monitor_08", "Replay", "Replay_01", "Replay_02", "Replay_03", "Replay_04", "Replay_05", "Replay_06", "Replay_07", "Replay_08", "AuthUserMag", "AuthSysCfg", "AuthSysInfo", "AuthManuCtr", "AuthStoreCfg", "AuthEventCfg", "AuthNetCfg", "AuthRmtDevice", "AuthSecurity", "AuthBackup", "AuthMaintence"],
                        "Sharable":"True",                    
                        "Reserved": "False",
                        "Type": "normal",
                        "Memo": "user1",
                        "MacOnly":"",
                        "MaxMonitorChannels": 0,
                    },
                },
            }
            

sending:
{'method': 'userManager.addUser', 'params': {'user': {'Name': 'user1', 'Password': '2E367142AB0650EEA0834544F7C1ECF7', 'Group': 'admin', 'AuthorityList': ['Monitor', 'Monitor_01', 'Monitor_02', 'Monitor_03', 'Monitor_04', 'Monitor_05', 'Monitor_06', 'Monitor_07', 'Monitor_08', 'Replay', 'Replay_01', 'Replay_02', 'Replay_03', 'Replay_04', 'Replay_05', 'Replay_06', 'Replay_07', 'Replay_08', 'AuthUserMag', 'AuthSysCfg', 'AuthSysInfo', 'AuthManuCtr', 'AuthStoreCfg', 'AuthEventCfg', 'AuthNetCfg', 'AuthRmtDevice', 'AuthSecurity', 'AuthBackup', 'AuthMaintence'], 'Sharable': 'True', 'Reserved': 'False', 'Type': 'normal', 'Memo': 'user1', 'MacOnly': '', 'MaxMonitorChannels': 0}}}   


but it still not login to web. what wrong?

**Reply by infiniband2** [reply]
user added, but password is wrong.
may be password store in another location?

{
    "Security": {
        "params": {
            "method": [
                "Security.getEncryptInfo",
                "Security.addUser",
                "Security.addUserPlain",
                "Security.modifyPasswordByType",
                "Security.getConfig",
                "Security.setConfig",
                "Security.getDefault",
                "Security.transferUKey",
                "Security.modifyPasswordPlain",
                "Security.listMethod"
            ]
        }
    }
}

**Reply by infiniband2** [reply]
{
             "AccessSchedule": [
                 ["1 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00"],
                 ["1 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00"],
                 ["1 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00"],
                 ["1 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00"],
                 ["1 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00"],
                 ["1 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00"],
                 ["1 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00", "0 00:00:00-24:00:00"]
             ],
             "AuthorityList": ["Monitor", "Monitor_01", "Monitor_02", "Monitor_03", "Monitor_04", "Monitor_05", "Monitor_06", "Monitor_07", "Monitor_08", "Replay", "Replay_01", "Replay_02", "Replay_03", "Replay_04", "Replay_05", "Replay_06", "Replay_07", "Replay_08", "AuthUserMag", "AuthSysCfg", "AuthSysInfo", "AuthManuCtr", "AuthStoreCfg", "AuthEventCfg", "AuthNetCfg", "AuthRmtDevice", "AuthSecurity", "AuthBackup", "AuthMaintence"],
             "Group": "admin",
             "Id": 11,
             "MacOnly": "",
             "MaxMonitorChannels": 0,
             "Memo": "user1",
             "Name": "user1",
             "Password": null,
             "PasswordModifiedTime": "2000:00:00-00:00:00",
             "Reserved": false,
             "Sharable": true,
             "Type": "Normal"
         }]
     },
     
     
     

in result i see that password is **NULL** ??? is it ok?

**Reply by infiniband2** [reply]
All ok! it's work. Problem in realm  ))) after debugging it's solved !!!  ))

**Reply by bakkano1** [reply]
Hi could you share your code for adding user.

**Reply by bakkano1** [reply]
> All ok! it's work. Problem in realm ))) after debugging it's solved !!! ))

Hello @infiniband2 would you be able to assist I am testing the dahua console on my home nvr and not sure how to run the code. thanks

**Reply by bakkano1** [reply]
what file do you guys add the code to
